[ADP-3306] Make `NoSuchWallet` error machine-readable.

paweljakubas opened the pull request:
- [x] Extend `NoSuchWallet` to include `WalletId`.
- [x] Adjust unit tests and swagger accordingly.
- [x] use `NoSuchWallet` rather than `err403NoSuchErrorMsg` throughout integration tests   

### Issue
ADP-3306

paweljakubas committed:
introduce ApiErrorNoSuchWallet error

paweljakubas committed:
swagger

paweljakubas committed:
remove err403NoSuchWallet - part 1

fix

paweljakubas committed:
refactor Shared.Wallet

paweljakubas committed:
refactor Shelley.StakePools

paweljakubas committed:
refactor Shared.Transactions

paweljakubas committed:
refactor Shelley.TransactionsNew

fix

paweljakubas committed:
Shelley.Addresses

paweljakubas committed:
refactor Byron.Migrations

paweljakubas committed:
Shelley.Migrations

paweljakubas committed:
refactor Shelley.Transaction

paweljakubas committed:
sort out DSL and ApiError naming overlap

paweljakubas committed:
Shelley.Wallets

paweljakubas committed:
Byron.Transactions

paweljakubas committed:
Byron.Wallets

paweljakubas committed:
Byron.Addresses

paweljakubas committed:
Byron.CoinSelections

paweljakubas committed:
revert CLI changes

jonathanknowles commented on lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -2378,7 +2381,9 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
-        expectErrorMessage (errMsg404NoWallet $ w ^. walletId) r
+        decodeErrorInfo r `shouldBe`
+            (NoSuchWallet $ ApiErrorNoSuchWallet $ w ^. walletId)
+
```
```suggestion
```

paweljakubas replied:
fixed

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types/Error.hs:
```diff
@@ -159,6 +160,8 @@ data ApiErrorInfo
     | NoSuchTransaction
     | NoSuchWallet
+        !ApiErrorNoSuchWallet
+    | NoSuchWalletInitialized
```
If this error is a **_known_** wallet that is simply not initialized, then could we use a name like:
```suggestion
    | WalletNotInitialized
```

paweljakubas replied:
sounds reasonable - done

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types/Error.hs:
```diff
@@ -312,3 +315,10 @@ data ApiErrorNoSuchPool = ApiErrorNoSuchPool
     deriving anyclass NFData
+
+data ApiErrorNoSuchWallet = ApiErrorNoSuchWallet
+    { walletId :: !Text
```
Suggestion: does it make sense to use `ApiT WalletId` here?

```suggestion
    { walletId :: !(ApiT WalletId)
```

paweljakubas replied:
trying to use pattern synonyms here -> https://github.com/cardano-foundation/cardano-wallet/pull/4597/commits/984bb6eb3c841acaefb22251c42f58a0b7327462

jonathanknowles commented on lib/integration/scenarios/Test/Integration/Scenario/API/Byron/Addresses.hs:
```diff
@@ -190,8 +195,9 @@ scenario_ADDRESS_LIST_04 fixture = it title $ \ctx -> runResourceT $ do
-        , expectErrorMessage $ errMsg404NoWallet $ w ^. walletId
         ]
+    decodeErrorInfo r `Lifted.shouldBe`
+        (NoSuchWallet $ ApiErrorNoSuchWallet $ w ^. walletId)
```
For cases like this, I would suggest:

```suggestion
        NoSuchWallet (ApiErrorNoSuchWallet $ w ^. walletId)
```

jonathanknowles replied:
(I think `hlint` highlights a series of similar cases.)

paweljakubas replied:
indeed, done

jonathanknowles replied:
@paweljakubas `hlint` is also failing for these: https://buildkite.com/cardano-foundation/cardano-wallet/builds/4956#018f8702-7a5f-465b-8b49-1b376461a17a

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -4361,6 +4361,18 @@ x-errNoUtxosAvailable: &errNoUtxosAvailable
       type: string
       enum: ['no_utxos_available']
 
+x-errNoSuchWalletInitialized: &errNoSuchWalletInitialized
```
Related to https://github.com/cardano-foundation/cardano-wallet/pull/4597#discussion_r1606488407 (above):

```suggestion
x-errWalletNotInitialized: &errWalletNotInitialized
```

paweljakubas replied:
done

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -4361,6 +4361,18 @@ x-errNoUtxosAvailable: &errNoUtxosAvailable
 
+x-errNoSuchWalletInitialized: &errNoSuchWalletInitialized
+  <<: *responsesErr
+  title: no_such_wallet_initialized
```
Related to https://github.com/cardano-foundation/cardano-wallet/pull/4597#discussion_r1606488407 (above):

```suggestion
  title: wallet_not_initialized
```

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -5627,6 +5645,13 @@ x-responsesErr404WalletNotFound: &responsesErr404WalletNotFound
       application/json:
         schema: *errNoSuchWallet
 
+x-responsesErr404DbNotInitialized: &responsesErr404DbNotInitialized
```
Related to https://github.com/cardano-foundation/cardano-wallet/pull/4597#discussion_r1606488407 (above):

How about:
```suggestion
x-responsesErr404WalletNotInitialized: &responsesErr404WalletNotInitialized
```

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -5627,6 +5645,13 @@ x-responsesErr404WalletNotFound: &responsesErr404WalletNotFound
 
+x-responsesErr404DbNotInitialized: &responsesErr404DbNotInitialized
+  404:
+    description: DB uninitialized
```
```suggestion
    description: Wallet not yet initialized.
```

jonathanknowles reviewed (commented):
Hi @paweljakubas 

Thanks for making this PR! I've left a few suggestions.

I believe the unit test failure will be fixed by: https://github.com/cardano-foundation/cardano-wallet/pull/4598.

One last request: would it be possible to fill out the PR description?

Many thanks again!

paweljakubas committed:
reqquested review changes

jonathanknowles committed:
Make `WalletId` an instance of `Data`.

jonathanknowles committed:
Use `ApiT WalletId` instead of `Text` for `ApiErrorNoSuchWallet`.

jonathanknowles committed:
Regenerate `ApiErrorNoSuchWallet.json`.

This is necessary, as type `ApiErrorNoSuchWallet` uses a real `WalletId`
instead of just `Text`.

jonathanknowles committed:
Rename `Err404DbNotInitialized` to `Err404WalletNotInitialized`.

jonathanknowles reviewed (approved):
LGTM!

@paweljakubas, many thanks for making this PR!

jonathanknowles committed:
Merge branch 'master' into paweljakubas/adp-3306/refactor-no-such-wallet

jonathanknowles committed:
Fix order of imports, placed out of order by prior merge.

jonathanknowles committed:
Revert use of `decodeErrorInfo` for `WALLETS_SIGNATURES_03`.

This particular test happens to use `rawRequest` instead of `request`.
Whereas the `request` function attempts to decode the response into a
JSON `Value`, the `rawRequest` function does not perform any decoding.

This causes the subsequent `decodeErrorInfo` to fail with:

```
  uncaught exception: ErrorCall
   decodeErrorInfo: Expected a 'ClientError', but encountered something else: Left (RawClientError "{\"code\":\"no_such_wallet\",\"info\":{\"wallet_id\":\"acab1951ad1ef5808eb9756b02d1ff85caf0e702\"},\"message\":\"I couldn't find a wallet with the given id: acab1951ad1ef5808eb9756b02d1ff85caf0e702\"}")
   CallStack (from HasCallStack):
     error, called at framework/Test/Integration/Framework/DSL.hs:742:13 in cardano-wallet-integration-2024.5.5-KhWmuWPIqBp7pTemTBE3I2-framework:Test.Integration.Framework.DSL
     decodeErrorInfo, called at scenarios/Test/Integration/Scenario/API/Shelley/Wallets.hs:1495:9 in cardano-wallet-integration-2024.5.5-AcHGWrd1jBGKoKhY01VJRG-scenarios:Test.Integration.Scenario.API.Shelley.Wallets
```
